README.md
# WhatScheduler by **DreadedHippy**
![GitHub last commit](https://img.shields.io/github/last-commit/DreadedHippy/WhatScheduler_backend?color=%23ffbb00&logo=Github&logoColor=%23ffbb00&style=for-the-badge)
![GitHub package.json version](https://img.shields.io/github/package-json/v/DreadedHippy/WhatScheduler_backend?color=%23ffbb00&logo=Github&logoColor=%23ffdd00&style=for-the-badge)

##### This is a simple NodeJS program to automate sending and replying to whatsapp messages using *node-cron*,*qrcode-terminal* and *whatsapp-web* NPM packages. I hope to implement scheduled job-postings using Puppeteer. Buckle up for that 😉

#### My motivation behind this project:
Over the long period of time I've been using whatsapp (3+ Years), there were moments when I forgot
to send messages such as birthday messages or messages requesting permission from someone and I thought: I've been coding since 2019, why don't I just *automate* it? Here we are 😅


#### [ChangeLogs](CHANGELOG.md)

#### How to Install and run:
*Disclaimer: This was made with Node.js v18.12.1*
- Clone the repository
- Run these commands in the terminal: `npm i` and then `npm run start`
- You're all set

#### Using the project:
When you authenticate for the first time, you are required to login by scanning the QR code using *Whatsapp* installed on your phone.
***Note*: This is the default configuration, you can check out more authentication methods with the *whatsapp-web.js* package at [Whatsapp-Web website]('https://wwebjs.dev/guide/')**

After the first authentication, a folder named '.wwebjs_auth' will be created in the pwd, which will store session information. Basically, you only have to login *once*.

By default, while the program is running, it sends the message `Hello from the other side!` every 30 secondss. This was done using with the help of *node-cron* package,
**More information at [node-cron repo page]('https://github.com/node-cron/node-cron')**


#### Errors and solutions:
| Error | Possible cause | Solution |
|----|------|--------|
| Could not find expected browser (chrome) locally. Run `npm install` to download the correct Chromium revision (982053).| The program is unable to detect a browser usable by the puppeteer package dependency of whatsapp-web.js package| Run the command `node ./node_modules/whatsapp-web.js/node_modules/puppeteer/install.js`|


#### License:
MIT License

Copyright (c) 2022 Onotioese Izormen

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
## v1.0.1 (2022-12-10)

### Highlights
- Added Job Scraping from Linkedin


## v1.1.0 (2023-01-11)

### Highlights
#### Removed:
- Job Scraping from Linkedin

#### Added
+ Connection to frontend
+ Resolved issue with whatsapp-web.js
	 + [View Resolution](https://github.com/pedroslopez/whatsapp-web.js/issues/1913#issuecomment-1377194341)
+ Message sending from frontend
+ Message scheduling from frontend

scraper.js
const axios = require('axios');
const cheerio = require('cheerio');
const ObjectsToCsv = require('objects-to-csv');

linkedinJobs = [];
let pageNumber = 0;
    
function getInfo() {
    return new Promise((resolve, reject) => {
    setTimeout(() => {
        let url = `https://www.linkedin.com/jobs/search?keywords=Web%20Development&location=Lagos%2C%20Lagos%20State%2C%20Nigeria&geoId=105693087&trk=public_jobs_jobs-search-bar_search-submit&position=1&pageNum=0&start=${pageNumber}`;
        axios(url).then(response => {
            const html = response.data;
            const $ = cheerio.load(html);
            const jobs = $('li')
            jobs.each((index, element) => {
                const jobTitle = $(element).find('h3.base-search-card__title').text().trim()
                const company = $(element).find('h4.base-search-card__subtitle').text().trim()
                const location = $(element).find('span.job-search-card__location').text().trim()
                const link = $(element).find('a.base-card__full-link').attr('href')
    
                linkedinJobs.push({
                    'Title': jobTitle,
                    'Company': company,
                    'Location': location,
                    'Link': link,
                })
    
                if(!jobTitle || !company || !location || !link){
                    linkedinJobs.pop()
                }
                resolve(linkedinJobs)
            });
            // const csv = new ObjectsToCsv(linkedinJobs)
            // csv.toDisk('./linkedInJobs.csv', { append: true })
        })
        .catch(console.error)

        pageNumber+= 25;
        if(pageNumber < 50){
            getInfo()
        }
    }, 5000);
})
}


let jobs = new Promise(function showResult(resolve, reject){
    getInfo().then((result) =>{
        resolve(result)
    }).catch(err => console.log(err))
})



module.exports = {jobs}



backend/app.mjs
import qrcode from 'qrcode-terminal'
import wweb from 'whatsapp-web.js';
const { Client, LocalAuth } = wweb
import schedule from 'node-schedule';
import nodeCron from 'node-cron';
// import scraper from './tools/scraper.js';
import dotenv from 'dotenv'
dotenv.config();
import express from 'express';
import cors from 'cors'
const app = express();
import bodyParser from 'body-parser'
// const index = require("../index")
let myGroup = {}
let isReady = false;


const mainGroupName = process.env.MAIN_GROUP_NAME
const testGroupName = 'Whatsapp'



// Use the saved values
const client = new Client({
	authStrategy: new LocalAuth()
});

client.on('qr', (qr) => {
    qrcode.generate(qr, {small: true});
});

client.on('ready', () => {
	console.log('Client is ready!');
	isReady = true
	client.getChats().then( chats => {
		// console.log(chats)
		myGroup = chats.find((chat) => chat.name === testGroupName);
		client.sendMessage(
			myGroup.id._serialized, 'Hello from the other side!'
		)

		let position = 0
	})
});

client.on('message', message => {
	console.log(message.body);
});

client.initialize()



app.use(cors());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false}));

app.get('/', (req, res) => {
  res.send('Hello World');
});

app.post('/simplesend', (req, res, next) => {
	const message = req.body
	const date = new Date

	function sendMessage(){
		client.sendMessage(
			myGroup.id._serialized, message.content
		).then(() => {
			console.log('Message sent!', 'Is instant?', message.isInstant)
			res.status(200).json({
				message: 'Sent!',
			})
		}).catch(err => {
			console.log('Message not sent!',err)					
			res.status(400).json({
				message: 'Something went wrong...',
			})
		})
	}

	if(!message.isInstant){
		const job = schedule.scheduleJob(message.date, function(){
			client.sendMessage(
				myGroup.id._serialized, message.content
			)
		})
		console.log('Scheduled message')
		res.status(200).json({
			message: 'Message Scheduled!',
		})
		return;
	}

	sendMessage()

})

function getState(){
	// console.log(isReady)
	return isReady
}

export default {getState, app};

index.js
const qrcode = require('qrcode-terminal')
const { Client, LocalAuth } = require('whatsapp-web.js');
const nodeCron = require('node-cron');
const scraper = require('./backend/tools/scraper');
require('dotenv').config()


let isReady = false
const mainGroupName = process.env.MAIN_GROUP_NAME
const testGroupName = 'Whatsapp'

// Use the saved values
const client = new Client({
    authStrategy: new LocalAuth()
});

client.on('qr', (qr) => {
    qrcode.generate(qr, {small: true});
});

client.on('ready', () => {
	console.log('Client is ready!');
	isReady = true
	client.getChats().then( chats => {
		const myGroup = chats.find((chat) => chat.name === testGroupName);
		client.sendMessage(
			myGroup.id._serialized, 'Hello from the other side!'
		)

		let position = 0

		scraper.jobs.then((result) => {
			const job = nodeCron.schedule("*/30 * * * * *", () => {
				let title = result[position].Title
				let company = result[position].Company
				let location = result[position].Location
				let link = result[position].Link

				if(position > 50){
					return client.sendMessage(
						myGroup.id._serialized, `Hello, this is a scheduled message(position ${position})!`
					)
				}
				
				client.sendMessage(
					myGroup.id._serialized, ` 💡\`\`\`LINKEDIN JOB ALERTS\`\`\`💡
					\n *Title*: ${title}
					\n *Company*: ${company}
					\n *Location*: ${location}
					\n *Link*: ${link}
					`
				)
				position++
			});
		}).catch( err => {console.log(err)})

	})
});

client.on('message', message => {
	console.log(message.body);
});

function sendMessage(message){
	client.sendMessage(
		myGroup.id._serialized, ` 💡\`\`\`LINKEDIN JOB ALERTS\`\`\`💡
		\n *Title*: ${title}
		\n *Company*: ${company}
		\n *Location*: ${location}
		\n *Link*: ${link}
		`
	)
}

client.initialize();
